### PR TITLE
WIP: Sid/rlp constraints

### PR DESCRIPTION
**Pull request by siddhantsharma301**
Working on: https://github.com/succinctlabs/succinctx/issues/58

### CODE REVIEW

File: plonky2x/src/frontend/eth/mpt/builder.rs [hunk 1]
```diff
@@ -51,7 +51,26 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
             _phantom: PhantomData,
         };
         self.add_simple_generator(generator.clone());
-        generator.output
+
```
Comment (puma314) [inline plonky2x/src/frontend/eth/mpt/builder.rs:54]: nit: maybe we should have 2 functions `lt` and `le` and basically `le` will call `lt` and then at the end do 
```
        let eq = self.is_equal(lhs, rhs);
        let lte = self.or(lt, eq);
```

File: plonky2x/src/frontend/eth/mpt/builder.rs [hunk 2]
```diff
@@ -51,7 +51,26 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
...
+        // typically stored in BE, but we can just index instead of reversing
+        let res_bits = self.api.split_le(res.0, NUM_LIMBS);
+        let last_bit = Variable::from(res_bits[NUM_LIMBS - 1].target);
+        let lt_var = self.sub(_one, last_bit);
+        let lt = BoolVariable::from(lt_var);
+        let eq = self.is_equal(lhs, rhs);
+        let lte = self.or(lt, eq);
+
+        let output = generator.output;
+        self.assert_is_equal(lte, output);
```
Comment (puma314) [inline plonky2x/src/frontend/eth/mpt/builder.rs:72]: In this case, we should just return `lte` and remove the generator

File: plonky2x/src/frontend/eth/mpt/builder.rs (hunk 1 above)
Comment (puma314) [inline plonky2x/src/frontend/eth/mpt/builder.rs:54]: Would be great to have a bunch of test cases for this testing the edge cases and also overflow/underflow situations

File: plonky2x/src/frontend/eth/mpt/builder.rs [hunk 3]
```diff
@@ -51,7 +51,26 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
...
+        output
     }
 
     pub fn byte_to_variable(&mut self, lhs: ByteVariable) -> Variable {
```
[the remark this answers is not in the document]
Comment (siddhantsharma301) [reply plonky2x/src/frontend/eth/mpt/builder.rs]: done in https://github.com/succinctlabs/succinctx/pull/152/commits/2e9d50bfe18d63fb9104f4d0db930ac9259ef19c

File: plonky2x/src/frontend/eth/mpt/builder.rs [hunk 4]
```diff
@@ -61,7 +80,20 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
...
+        let _two = self.constant::<Variable>(L::Field::from_canonical_u8(2));
+        let mut acc = self.zero::<Variable>();
+        let mut pow = self.one::<Variable>();
+        for i in 0..8 {
+            let curr = lhs.0[i].0;
+            let tmp = self.mul(acc, pow);
+            acc = self.add(tmp, curr);
+            pow = self.mul(pow, _two);
+        }
+        self.assert_is_equal(output, acc);
```
[the remark this answers is not in the document]
Comment (siddhantsharma301) [reply plonky2x/src/frontend/eth/mpt/builder.rs]: done in https://github.com/succinctlabs/succinctx/pull/152/commits/2e9d50bfe18d63fb9104f4d0db930ac9259ef19c

File: plonky2x/src/frontend/vars/byte.rs [hunk 5]
```diff
@@ -237,6 +237,23 @@ impl<L: PlonkParameters<D>, const D: usize> Zero<L, D> for ByteVariable {
     }
 }
 
+impl<L: PlonkParameters<D>, const D: usize> Add<L, D> for ByteVariable {
+    type Output = Self;
+
+    fn add(self, rhs: Self, builder: &mut CircuitBuilder<L, D>) -> Self::Output {
```
[the remark this answers is not in the document]
Comment (siddhantsharma301) [reply plonky2x/src/frontend/vars/byte.rs:262]: added test in https://github.com/succinctlabs/succinctx/pull/152/commits/2f72c171c35a9adec47b0fa9f54b5784c9763c2a

File: plonky2x/src/frontend/eth/mpt/builder.rs (hunk 1 above)
Comment (siddhantsharma301) [reply plonky2x/src/frontend/eth/mpt/builder.rs]: added general test in https://github.com/succinctlabs/succinctx/pull/152/commits/b79b4e7d2d4fc49e75e4aced94cddf1a78a0a9c7, need to add specific edge case tests

File: plonky2x/src/frontend/eth/mpt/builder.rs [hunk 6]
```diff
@@ -51,28 +51,26 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
             _phantom: PhantomData,
         };
         self.add_simple_generator(generator.clone());
```
[the remark this answers is not in the document]
Comment (siddhantsharma301) [reply plonky2x/src/frontend/eth/mpt/builder.rs]: fixed in https://github.com/succinctlabs/succinctx/pull/152/commits/105ccb5087670c4af0c5749a92639e84e88c73d9

File: plonky2x/src/frontend/eth/mpt/builder.rs (hunk 2 above)
Comment (siddhantsharma301) [reply plonky2x/src/frontend/eth/mpt/builder.rs]: fixed in https://github.com/succinctlabs/succinctx/pull/152/commits/105ccb5087670c4af0c5749a92639e84e88c73d9

File: plonky2x/src/frontend/eth/mpt/builder.rs (hunk 1 above)
Comment (siddhantsharma301) [reply plonky2x/src/frontend/eth/mpt/builder.rs]: fixed

Comment (siddhantsharma301) [reply plonky2x/src/frontend/eth/mpt/builder.rs]: added some edge case test in https://github.com/succinctlabs/succinctx/pull/152/commits/da43273874a36ba41bf2ba3d333211c7232107d2

File: plonky2x/src/frontend/vars/byte.rs [hunk 7]
```diff
@@ -118,6 +121,22 @@ impl ByteVariable {
 
         [ByteVariable(left_nibble), ByteVariable(right_nibble)]
     }
+
+    pub fn to_variable<L: PlonkParameters<D>, const D: usize>(
+        self,
+        builder: &mut CircuitBuilder<L, D>,
+    ) -> Variable {
+        let _two = builder.constant::<Variable>(L::Field::from_canonical_u8(2));
+        let mut acc = builder.zero::<Variable>();
```
Comment (puma314) [inline plonky2x/src/frontend/vars/byte.rs:130]: @jtguibas is this the preferred way for getting `zero` and `one` as constants? I'm not sure

File: plonky2x/src/frontend/vars/byte.rs [hunk 8]
```diff
@@ -237,6 +256,43 @@ impl<L: PlonkParameters<D>, const D: usize> Zero<L, D> for ByteVariable {
     }
 }
 
+impl<L: PlonkParameters<D>, const D: usize> Add<L, D> for ByteVariable {
+    type Output = Self;
+
+    fn add(self, rhs: Self, builder: &mut CircuitBuilder<L, D>) -> Self::Output {
+        let mut bits = array![_ => builder.init::<BoolVariable>(); 8];
```
Comment (puma314) [inline plonky2x/src/frontend/vars/byte.rs:263]: I believe here, you want to use `builder.init_unsafe` (which is something John just added), since you're going to be just setting elements of `bits` from safe variables. @jtguibas can you check me on this?

File: plonky2x/src/frontend/vars/variable.rs [hunk 9]
```diff
@@ -107,3 +109,142 @@ impl<L: PlonkParameters<D>, const D: usize> One<L, D> for Variable {
...
 }
+
+impl<L: PlonkParameters<D>, const D: usize> LessThan<L, D> for Variable {
+    type Output = BoolVariable;
+
+    fn lt(self, rhs: Variable, builder: &mut CircuitBuilder<L, D>) -> Self::Output {
+        // FIXME: Check for overflows and edge cases
+        const NUM_LIMBS: usize = 64;
+        let max = builder.constant(L::Field::from_canonical_u64(1 << (NUM_LIMBS - 1)));
+        let _one: Variable = builder.constant(L::Field::from_canonical_u64(1));
```
Comment (puma314) [inline plonky2x/src/frontend/vars/variable.rs:120]: nit: I think you should just do `builder.one::<Variable>()` here

File: plonky2x/src/frontend/vars/variable.rs [hunk 10]
```diff
@@ -107,3 +109,142 @@ impl<L: PlonkParameters<D>, const D: usize> One<L, D> for Variable {
         Variable(builder.api.one())
     }
 }
+
+impl<L: PlonkParameters<D>, const D: usize> LessThan<L, D> for Variable {
+    type Output = BoolVariable;
+
+    fn lt(self, rhs: Variable, builder: &mut CircuitBuilder<L, D>) -> Self::Output {
+        // FIXME: Check for overflows and edge cases
```
Comment (puma314) [inline plonky2x/src/frontend/vars/variable.rs:117]: what are the overflows and edge cases here?

File: plonky2x/src/frontend/vars/variable.rs [hunk 11]
```diff
@@ -107,3 +109,142 @@ impl<L: PlonkParameters<D>, const D: usize> One<L, D> for Variable {
...
+        let max = builder.constant(L::Field::from_canonical_u64(1 << (NUM_LIMBS - 1)));
+        let _one: Variable = builder.constant(L::Field::from_canonical_u64(1));
+        // from circomlib: https://github.com/iden3/circomlib/blob/master/circuits/comparators.circom#L89
+        let tmp = builder.add(self, max);
+        let res = builder.sub(tmp, rhs);
+        // typically stored in BE, but we can just index instead of reversing
+        let res_bits = builder.api.split_le(res.0, NUM_LIMBS);
+        let last_bit = Variable::from(res_bits[NUM_LIMBS - 1].target);
+        let lt_var = builder.sub(_one, last_bit);
+        BoolVariable::from(lt_var)
```
Comment (puma314) [inline plonky2x/src/frontend/vars/variable.rs:128]: Do we need to do a range_check here when initializing the bool variable?

File: plonky2x/src/frontend/vars/variable.rs [hunk 12]
```diff
@@ -107,3 +109,142 @@ impl<L: PlonkParameters<D>, const D: usize> One<L, D> for Variable {
         Variable(builder.api.one())
     }
 }
+
+impl<L: PlonkParameters<D>, const D: usize> LessThan<L, D> for Variable {
+    type Output = BoolVariable;
+
+    fn lt(self, rhs: Variable, builder: &mut CircuitBuilder<L, D>) -> Self::Output {
+        // FIXME: Check for overflows and edge cases
+        const NUM_LIMBS: usize = 64;
```
Comment (puma314) [inline plonky2x/src/frontend/vars/variable.rs:118]: Shouldn't this depend on the Field? I think in particular `F::order()` should work and return to you the order of the field, and you can get `NUM_LIMBS` from that. 

Another option is that here you panic if `2**{NUM_LIMBS} != log_2(order)` just as a runtime check since I think it's quite unlikely that people will not use Goldilocks (there might be an off by one in the equation above but you get the point).

File: plonky2x/src/frontend/vars/variable.rs [hunk 13]
```diff
@@ -107,3 +109,142 @@ impl<L: PlonkParameters<D>, const D: usize> One<L, D> for Variable {
...
+    fn test_lt() {
+        let mut builder = CircuitBuilder::<L, D>::new();
+        let lhs = builder.read::<Variable>();
+        let rhs = builder.read::<Variable>();
+        let less_than = builder.lt(lhs, rhs);
+        builder.write(less_than);
+
+        let circuit = builder.build();
+
+        let rand_lhs = rand::random::<u64>();
```
Comment (puma314) [inline plonky2x/src/frontend/vars/variable.rs:161]: Won't this fail if `rand_lhs` is > the field order? or there are some edge cases here I think that we shoudl be careful of

**Review by puma314 — COMMENTED** [review]
overall very close and thanks for adding all the tests! looks v clean, just left some smaller nit comments regarding some new apis that we added for init_safe and init_unsafe

**Review by puma314 — APPROVED** [review]

File: plonky2x/src/frontend/vars/variable.rs (hunk 13 above)
Comment (siddhantsharma301) [reply plonky2x/src/frontend/vars/variable.rs:161]: fixed in [c5c901d](https://github.com/succinctlabs/succinctx/pull/152/commits/c5c901d1da6ce4cb7d2237f81266d01e054632d0)

File: plonky2x/src/frontend/vars/variable.rs (hunk 9 above)
Comment (siddhantsharma301) [reply plonky2x/src/frontend/vars/variable.rs]: fixed in [96d77c3](https://github.com/succinctlabs/succinctx/pull/152/commits/96d77c370e478a2372630da59b9cc996b23e4a0a)

File: plonky2x/src/frontend/vars/variable.rs (hunk 12 above)
Comment (siddhantsharma301) [reply plonky2x/src/frontend/vars/variable.rs]: fixed here [4d760f8](https://github.com/succinctlabs/succinctx/pull/152/commits/4d760f819628fff3fb5158ae8b6938ce6e607bfc)
